libs/brig-types/src/Brig/Types/User.hs

```diff
@@ -229,41 +243,82 @@ data NewUser = NewUser
     , newUserPassword       :: !(Maybe PlainTextPassword)
-    , newUserTeam           :: !(Maybe NewTeamUser)
-    , newUserExpiresIn      :: !(Maybe (Range 1 604800 Integer)) -- ^ 1 second - 1 week
+    , newUserExpiresIn      :: !(Maybe ExpiresIn)
     }
+    deriving (Eq, Show)
+
+-- | 1 second - 1 week
+type ExpiresIn = Range 1 604800 Integer
+
+data NewUserOrigin =
+    NewUserOriginInvitationCode !InvitationCode
+  | NewUserOriginTeamUser !NewTeamUser
+  deriving (Eq, Show)
+
+parseNewUserOrigin :: Maybe PlainTextPassword -> Maybe UserIdentity -> Maybe UserSSOId
+                   -> Object -> Parser (Maybe NewUserOrigin)
+parseNewUserOrigin pass uid ssoid o = do
+    invcode  <- o .:? "invitation_code"
+    teamcode <- o .:? "team_code"
+    team     <- o .:? "team"
+    teamid   <- o .:? "team_id"
+    result <- case (invcode, teamcode, team, ssoid, teamid) of
+        (Just a,  Nothing, Nothing, Nothing, Nothing) -> return . Just . NewUserOriginInvitationCode $ a
+        (Nothing, Just a,  Nothing, Nothing, Nothing) -> return . Just . NewUserOriginTeamUser $ NewTeamMember a
+        (Nothing, Nothing, Just a,  Nothing, Nothing) -> return . Just . NewUserOriginTeamUser $ NewTeamCreator a
+        (Nothing, Nothing, Nothing, Just _,  Just t)  -> return . Just . NewUserOriginTeamUser $ NewTeamMemberSSO t
+        (Nothing, Nothing, Nothing, Nothing, Nothing) -> return Nothing
+        (_, _, _, _, _) -> fail $ "team_code, team, invitation_code, sso_id are mutually exclusive\
+                                  \ and sso_id, team_id must be either both present or both absent."
+    case (result, pass, uid) of
+        (_, _, Just SSOIdentity {}) -> pure result
+        (Just (NewUserOriginTeamUser _), Nothing, _) -> fail "all team users must set a password on creation"
+        _ -> pure result
+
+jsonNewUserOrigin :: NewUserOrigin -> [Pair]
+jsonNewUserOrigin = \case
+    NewUserOriginInvitationCode inv             -> ["invitation_code" .= inv]
+    NewUserOriginTeamUser (NewTeamMember tc)    -> ["team_code" .= tc]
+    NewUserOriginTeamUser (NewTeamCreator team) -> ["team" .= team]
+    NewUserOriginTeamUser (NewTeamMemberSSO ti) -> ["team_id" .= ti]
+
+newUserInvitationCode :: NewUser -> Maybe InvitationCode
+newUserInvitationCode nu = case newUserOrigin nu of
+    Just (NewUserOriginInvitationCode ic) -> Just ic
+    _ -> Nothing
+
+newUserTeam :: NewUser -> Maybe NewTeamUser
+newUserTeam nu = case newUserOrigin nu of
+    Just (NewUserOriginTeamUser tu) -> Just tu
+    _ -> Nothing
 
 newUserEmail :: NewUser -> Maybe Email
 newUserEmail = emailIdentity <=< newUserIdentity
 
 newUserPhone :: NewUser -> Maybe Phone
 newUserPhone = phoneIdentity <=< newUserIdentity
 
+newUserSSOId :: NewUser -> Maybe UserSSOId
+newUserSSOId = ssoIdentity <=< newUserIdentity
+
 instance FromJSON NewUser where
       parseJSON = withObject "new-user" $ \o -> do
+          ssoid                 <- o .:? "sso_id"
```

Comment:
Any reason not to parse `sso_id` in `parseIdentity` ?

Reply:
we could have parsed it in both `parseIdentity` and `parseNewUserOrigin` (further down in the same do block).  but i like the fact that it's only parsed once, so the parser only needs to be changed in one place if it has to be changed.  and i thought it is clearer that way.

see also: https://github.com/wireapp/wire-server/blob/feature/sso-brig-integration/libs/brig-types/src/Brig/Types/User.hs#L164